### PR TITLE
feat: add walk-forward scoring backtest

### LINKED ISSUE #41: Add walk-forward backtesting for the scoring engine
## Context

The current scoring engine is a deterministic baseline that combines cross-sectional ranks for momentum, trend, volatility, drawdown, RSI, and z-score features. This is useful for reproducible reporting, but the repository does not yet measure whether high scores are associated with better subsequent returns.

## Scope

- Add a lightweight walk-forward backtesting command for saved OHLCV data.
- Evaluate whether top-ranked instruments outperform lower-ranked instruments over configurable forward horizons.
- Support at least daily horizons such as 1d, 5d, 20d, and 60d.
- Report basic metrics such as forward return by score bucket, top-k average return, hit rate, turnover, and drawdown where practical.
- Generate machine-readable backtest artifacts under a deterministic path such as `data/backtests/`.
- Keep the implementation dependency-light and deterministic.

## Non-goals

- Do not add broker execution, order simulation, leverage, financing costs, or portfolio optimization in this issue.
- Do not turn reports into investment advice.

## Acceptance criteria

- A CLI command can run walk-forward backtests from saved OHLCV data.
- Tests cover deterministic fixtures and edge cases such as insufficient history and missing symbols.
- Backtest artifacts include config, scoring version, date range, forward horizons, and metrics.
- `OPERATIONS.md` explains how to interpret backtest outputs and their limitations.
- Local QA passes via `.agents/skills/local-qa/scripts/qa.sh`.

### PR DESCRIPTION
**Pull request by dceoy**
## Summary

- add a deterministic walk-forward backtesting CLI for saved daily OHLCV data
- report forward returns by score bucket, top-k returns and hit rate, turnover, and drawdown
- write machine-readable artifacts under `data/backtests/`
- document usage, interpretation, and limitations

## Validation

- `.agents/skills/local-qa/scripts/qa.sh`
- 329 tests passed with 100% branch coverage

Closes #41

### CODE REVIEW

File: .agents/skills/market-analysis/scripts/backtest.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+            for s in score_instruments(
+                window, reference_time=date, min_history=min_history
+            )
+            if s.is_reliable
+        ]
+        eligible = [
+            s
+            for s in scores
+            if all(
+                positions[s.symbol][date] + h < len(data[s.symbol]) for h in horizons
```
[the remark this answers is not in the document]
Comment (dceoy) [reply .agents/skills/market-analysis/scripts/backtest.py]: Fixed in 15c271b. Eligibility and top-k selection are now computed independently for each horizon, so longer horizons no longer truncate shorter-horizon samples or one-day drawdown observations.

File: .agents/skills/market-analysis/scripts/backtest.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+    top_hits: dict[int, list[bool]] = defaultdict(list)
+    daily_top_returns: list[float] = []
+    turnovers: list[float] = []
+    previous: set[str] | None = None
+    observations = 0
+
+    for date in dates:
+        window = {s: bars[: positions[s][date] + 1] for s, bars in data.items()}
+        if any(len(bars) < min_history for bars in window.values()):
+            continue
```
[the remark this answers is not in the document]
Comment (dceoy) [reply .agents/skills/market-analysis/scripts/backtest.py]: Fixed in 15c271b. Each date now scores only symbols with sufficient history; a newly listed or otherwise short-history symbol no longer suppresses mature symbols. Added regression coverage for mixed history lengths.

**Review by dceoy — COMMENTED** [review]
Reviewed the latest commit. The previous Codex findings are resolved and CI is green, but I found one remaining backtest-validity issue that should be fixed before merge.

File: .agents/skills/market-analysis/scripts/backtest.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+                if positions[score.symbol][date] + horizon < len(data[score.symbol])
+            ]
+            if not eligible:
+                continue
+            date_observed = True
+            horizon_selected = eligible[:top_k]
+            horizon_returns = []
+            for rank_index, score in enumerate(eligible):
+                bucket = min(buckets, rank_index * buckets // len(eligible) + 1)
+                index = positions[score.symbol][date]
```
Comment (dceoy) [inline .agents/skills/market-analysis/scripts/backtest.py:94]: `horizon_selected = eligible[:top_k]` re-selects the portfolio after filtering by future availability. If a currently top-ranked symbol lacks `index + horizon`, the next lower-ranked symbol is promoted using information that is only known after the scoring date. This introduces availability/look-ahead bias and also makes the reported `turnover` inconsistent, because turnover is computed from `scores[:top_k]` before this horizon-specific re-selection. Prefer assigning buckets and top-k membership from the score ranking at the scoring date, then only aggregate forward returns for members that actually have a forward bar (or explicitly report missing forward observations).

Comment (dceoy) [reply .agents/skills/market-analysis/scripts/backtest.py:94]: Fixed in 26de427. Top-k and bucket membership now come exclusively from the score ranking at the scoring date. Missing forward bars are skipped without promoting lower-ranked symbols, and regression coverage verifies this behavior.

File: .agents/skills/market-analysis/scripts/backtest.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+            horizons=args.horizons,
+            top_k=args.top_k,
+            buckets=args.buckets,
+            min_history=args.min_history,
+        )
+    except (FileNotFoundError, ValueError) as exc:
+        parser.error(str(exc))
+    start = max(bars[0].timestamp for bars in data.values()).date().isoformat()
+    end = min(bars[-1].timestamp for bars in data.values()).date().isoformat()
+    artifact = {
```
Comment (dceoy) [inline .agents/skills/market-analysis/scripts/backtest.py:192]: The backtest now intentionally scores symbols independently when another requested symbol has short or missing history, but this metadata still reports the intersection of all input files (`max(first_bar)` to `min(last_bar)`). In a mixed-coverage universe, `run_backtest` can produce observations before that `start` or after that `end` using the mature symbols, while the artifact filename/date_range would say the run covered only the shorter common span. That makes the machine-readable artifact misleading and weakens the issue acceptance criterion that artifacts include the backtest date range. Derive `date_range` from the actually observed scoring/forward-return dates, or explicitly separate input coverage metadata from evaluated observation coverage.

**Review by dceoy — COMMENTED** [review]
Reviewed the latest head `26de427`. CI is green and the previously reported per-horizon/per-symbol availability issues are resolved. I found one remaining artifact correctness issue that should be fixed before merge.

File: .agents/skills/market-analysis/scripts/backtest.py (hunk 4 above)
Comment (dceoy) [reply .agents/skills/market-analysis/scripts/backtest.py:192]: Fixed in bae3e6a. The result now records the first and last scoring dates that produced forward-return observations, and the artifact filename uses that evaluated range. Runs with no observations fail without writing a misleading artifact; mixed-coverage and no-observation cases have regression tests.

File: .agents/skills/market-analysis/scripts/backtest.py [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+            if s.is_reliable
+        ]
+        if not scores:
+            continue
+        selected = scores[:top_k]
+        selected_symbols = {s.symbol for s in selected}
+        if previous is not None:
+            turnovers.append(
+                1.0
+                - len(previous & selected_symbols)
```
Comment (dceoy) [inline .agents/skills/market-analysis/scripts/backtest.py]: `turnovers` is appended before `date_observed` is known, and `previous` is updated even for scored dates that later produce no forward-return observations. With a long-only horizon such as `--horizons 60`, the final 60 scored dates have no evaluated returns, but they still contribute turnover, while `observations`, `date_range`, buckets, top-k returns, and drawdown exclude them. This makes the artifact mix different samples across metrics. Prefer computing/appending turnover only for dates that have at least one evaluated forward observation, or report a separate scoring-period turnover explicitly.

**Review by dceoy — COMMENTED** [review]
Reviewed the latest head `bae3e6a`. The earlier look-ahead and date-range issues appear resolved. I found one remaining sample-consistency issue in turnover aggregation that should be addressed before relying on the backtest artifact.

File: .agents/skills/market-analysis/scripts/backtest.py (hunk 5 above)
Comment (dceoy) [reply .agents/skills/market-analysis/scripts/backtest.py]: Fixed in ab7153b. Turnover is now calculated and state is advanced only when the scoring date has at least one evaluated forward observation, matching the observations and date-range sample. Added a regression where ranking changes only on an unevaluable terminal date.

**Review by dceoy — COMMENTED** [review]
Reviewed the latest head `ab7153b`. The prior availability, look-ahead, date-range, and terminal-turnover findings are resolved. I found one remaining edge case where top-k turnover can still use a different sample from top-k forward returns when only lower-ranked instruments have forward bars.

File: .agents/skills/market-analysis/scripts/backtest.py [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+        date_observed = False
+        for horizon in horizons:
+            horizon_returns = []
+            for rank_index, score in enumerate(scores):
+                index = positions[score.symbol][date]
+                if index + horizon >= len(data[score.symbol]):
+                    continue
+                date_observed = True
+                bucket = min(buckets, rank_index * buckets // len(scores) + 1)
+                close = data[score.symbol][index].close
```
Comment (dceoy) [inline .agents/skills/market-analysis/scripts/backtest.py:99]: `date_observed` is set when any scored instrument has a forward bar, but `turnover` is specifically a top-k metric. In a mixed-coverage universe, a date where the selected top-k instruments lack `index + horizon` but lower-ranked instruments still have forward bars will advance `previous` and append turnover even though `top_returns`/`top_hits` receive no top-k observation for that date. That leaves turnover on a broader sample than the top-k performance metrics. Consider tracking a separate `top_k_observed` flag and advancing turnover only when at least one selected instrument contributes a forward return, or explicitly reporting bucket-observation turnover separately from top-k turnover.

Comment (dceoy) [reply .agents/skills/market-analysis/scripts/backtest.py:99]: Fixed in ce76ff2. Turnover state now advances only when at least one contemporaneously selected top-k instrument contributes a requested forward return. Lower-ranked bucket observations no longer broaden the turnover sample; the mixed-coverage regression asserts the resulting transition average.

File: .agents/skills/market-analysis/scripts/backtest.py [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+    *,
+    horizons: tuple[int, ...] = DEFAULT_HORIZONS,
+    top_k: int = 3,
+    buckets: int = 4,
+    min_history: int = 60,
+) -> dict[str, Any]:
+    """Score each available date without look-ahead and aggregate forward returns."""
+    if top_k < 1 or buckets < 1 or min_history < 1 or not horizons:
+        msg = "top-k, buckets, min-history, and horizons must be positive"
+        raise ValueError(msg)
```
Comment (dceoy) [inline .agents/skills/market-analysis/scripts/backtest.py:50]: `horizons` is only checked for positivity, but the aggregation dictionaries are keyed by the horizon value while the loops iterate over every element. If a caller passes `--horizons 1,1` or calls `run_backtest(..., horizons=(1, 1))`, the same 1-day forward returns are appended twice into `bucket_returns[1]` and `top_returns[1]`, inflating `count` under a single `1d` metrics key. Please reject duplicate horizons or normalize them once before aggregation.

File: .agents/skills/market-analysis/scripts/backtest.py [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+    start = result["date_range"]["start"]
+    end = result["date_range"]["end"]
+    if start is None or end is None:
+        parser.error("no backtest observations for the requested configuration")
+    artifact = {
+        "schema_version": "1.0.0",
+        "scoring_version": SCORING_VERSION,
+        "generated_at": datetime.now(UTC).isoformat(),
+        "config": {
+            "symbols": list(symbols),
```
Comment (dceoy) [inline .agents/skills/market-analysis/scripts/backtest.py:196]: The issue scope and `OPERATIONS.md` describe deterministic backtest artifacts, but `generated_at` makes the JSON content differ for identical inputs. If byte-level reproducibility is part of the artifact contract, remove this field or inject/freeze it in a controlled way; otherwise, the documentation should avoid calling the artifact itself deterministic.

**Review by dceoy — COMMENTED** [review]
Reviewed latest commit ce76ff2. CI is green and the previously reported backtest-validity issues are resolved. I left two remaining comments: one correctness issue around duplicate horizons, and one reproducibility/docs alignment concern around generated_at.

File: .agents/skills/market-analysis/scripts/backtest.py (hunk 7 above)
Comment (dceoy) [reply .agents/skills/market-analysis/scripts/backtest.py:50]: Fixed in fed1e58. run_backtest now rejects duplicate horizons before aggregation, including direct API calls and CLI input through the existing ValueError handling. Added regression coverage for duplicate rejection.

File: .agents/skills/market-analysis/scripts/backtest.py (hunk 8 above)
Comment (dceoy) [reply .agents/skills/market-analysis/scripts/backtest.py:196]: Fixed in fed1e58. Removed the wall-clock generated_at field, so identical saved OHLCV inputs and configuration now produce byte-identical JSON. The CLI test runs the same configuration twice and verifies exact artifact equality.

File: .agents/skills/market-analysis/scripts/backtest.py [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+    *,
+    horizons: tuple[int, ...] = DEFAULT_HORIZONS,
+    top_k: int = 3,
+    buckets: int = 4,
+    min_history: int = 60,
+) -> dict[str, Any]:
+    """Score each available date without look-ahead and aggregate forward returns."""
+    if top_k < 1 or buckets < 1 or min_history < 1 or not horizons:
+        msg = "top-k, buckets, min-history, and horizons must be positive"
+        raise ValueError(msg)
```
Comment (dceoy) [inline .agents/skills/market-analysis/scripts/backtest.py:50]: `run_backtest()` still accepts non-positive horizon values when called directly, even though this guard and error message say horizons must be positive. The CLI parser rejects `--horizons 0`/negative values, but a direct call such as `run_backtest(data, horizons=(0,))` records zero-return observations, and `horizons=(-1,)` can read prior bars via negative indexing. Consider extending this validation to `any(h < 1 for h in horizons)` and adding a regression alongside the duplicate-horizon test.

**Review by dceoy — COMMENTED** [review]
Reviewed latest head `fed1e58`. CI is green and the earlier review threads are resolved. I left one minor correctness comment for the direct `run_backtest()` API path; the CLI path already rejects this input.

File: .agents/skills/market-analysis/scripts/backtest.py (hunk 9 above)
Comment (dceoy) [reply .agents/skills/market-analysis/scripts/backtest.py:50]: Fixed in 5c5ffaf. run_backtest now rejects every horizon below 1 before indexing, matching the CLI contract. Regression coverage includes both zero and negative direct-API horizons.

**Review by dceoy — COMMENTED** [review]
Reviewed latest head `5c5ffaf`. The previously reported availability, look-ahead, date-range, turnover, duplicate-horizon, and deterministic-artifact issues are resolved. CI is green, the added validation/regression coverage is adequate for the issue scope, and I did not find remaining blocking issues.

I could not submit this as an approval because GitHub does not allow approving your own pull request from this account context.

File: .agents/skills/market-analysis/scripts/backtest.py [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+            "top_k": args.top_k,
+            "buckets": args.buckets,
+            "min_history": args.min_history,
+        },
+        **result,
+    }
+    args.output_dir.mkdir(parents=True, exist_ok=True)
+    path = args.output_dir / f"{start}_{end}_{args.interval}.json"
+    path.write_text(json.dumps(artifact, indent=2, sort_keys=True) + "\n")
+    print(path)
```
Comment (dceoy) [inline .agents/skills/market-analysis/scripts/backtest.py:208]: The output filename is derived only from the evaluated date range and interval, while the artifact body allows materially different configurations (`symbols`, `forward_horizons`, `top_k`, `buckets`, `min_history`). Two runs over the same evaluated period—for example different horizons or top-k values—will write the same `START_END_d.json` path and silently replace the prior result. Since this CLI is meant to produce machine-readable backtest artifacts for configurable experiments, include a deterministic config identifier in the filename or create a run subdirectory keyed by the normalized config to avoid losing comparable backtest results.

**Review by dceoy — COMMENTED** [review]
Reviewed the merged PR head. CI is green and the previously reported review threads are resolved. I found one remaining artifact-management issue worth addressing in a follow-up because repeated runs with different configurations can overwrite each other.
